Allow file:read_file_info() to be called on I/O devices

andrenth opened the pull request:
This gives us the equivalent of `fstat(2)`. I've tested on Linux and Windows. On the latter I had to use the `GetFinalPathNameByHandleW()` function, which requires `_WIN32_WINNT > = 0x600`. In the PR, I'm simply returning `ENOTSUP` for lower values.

I was unsure if I should include the preloaded beam files in the PR.

Finally, this PR depends on #2212, but only for the tests on directories. The test can be removed and added back when/if that PR is merged.

michalmuskala commented on erts/preloaded/src/prim_file.erl:
```diff
@@ -598,20 +601,37 @@ read_link_info(Name, Opts) ->
         error:_ -> {error, badarg}
     end.
 
+adjust_times(FileInfo, TimeType) ->
```
Would it make sense to catch the `posix` `TimeType` here and skip all the conversion logic and copying of the `file_info` record?

andrenth replied:
Probably, yes, but since the original code doesn't do that I didn't want to make unrelated changes.

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -291,14 +291,14 @@ static int is_ignored_name(int name_length, const WCHAR *name) {
 }
 
-static int get_drive_number(const efile_path_t *path) {
+static int get_drive_number(const WCHAR *path, int length) {
```
I'd prefer that this and `is_executable_file` remain untouched, as `efile_path_t` is a promise that the path strictly follows the format used throughout the module. We've had a lot of issues with inconsistent path handling in the past so I'd like to follow convention whenever possible.

(These changes should be unnecessary if you repurpose `internal_read_link` as suggested earlier.)

andrenth replied:
Yeah, I've already taken these changes back. I'm rebuilding to test with your previous suggestions now. I'm not sure if it's cygwin that does that, but grass grows faster than OTP builds in it...

garazdawi replied:
> I'm not sure if it's cygwin that does that, but grass grows faster than OTP builds in it...

I believe that mingw is faster in our experience, though not by much. It is all of the bash scripts in various places that makes it slow. They make it easier to maintain the build system on windows, but a lot slower to build.

andrenth commented:
I've added two commits to address the comments above (remove the `_WIN32_WINNT` check, refactor internal_read_link and is_name_surrogate to take a `HANDLE`  (I've renamed the latter to is_handle_surrogate), and use it from efile_read_handle_info instead of calling `GetFinalPathNameByHandleW` directly).

andrenth commented:
Please let me know if it’s ok to rebase.

jhogberg commented:
Go ahead, I'll take another look at this and add it to our nightly tests on Monday.

andrenth commented:
Done :)

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -795,74 +858,41 @@ posix_errno_t efile_read_info(const efile_path_t *path, int follow_links, efile_
-        if(result->a_time == -EPOCH_DIFFERENCE) {
-            result->a_time = result->m_time;
-        }
+    build_file_info(&native_file_info, &path, is_link, result);
```
`&path` should be `path` here.
```suggestion
    build_file_info(&native_file_info, path, is_link, result);
```

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -716,39 +716,90 @@ static int is_executable_file(const efile_path_t *path) {
+    return 0;
+}
+
+static void build_file_info(BY_HANDLE_FILE_INFORMATION *native_file_info, efile_path_t *path, int is_link, efile_fileinfo_t *result) {
```
`path` isn't modified here, so it should be `const`.
```suggestion
static void build_file_info(BY_HANDLE_FILE_INFORMATION *native_file_info, const efile_path_t *path, int is_link, efile_fileinfo_t *result) {
```

andrenth commented:
I've applied the fixes and rebased.

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -716,39 +716,90 @@ static int is_executable_file(const efile_path_t *path) {
+    }
 }
 
 posix_errno_t efile_read_info(const efile_path_t *path, int follow_links, efile_fileinfo_t *result) {
```
`attributes` need to be used for the root path fallbacks to work.

```diff
-        attributes = FILE_ATTRIBUTE_DIRECTORY;
+        native_file_info.dwFileAttributes = FILE_ATTRIBUTE_DIRECTORY;
     } else if(is_path_root(path)) {
         /* Local (or mounted) roots can be queried with GetFileAttributesW but
          * lack support for GetFileInformationByHandle, so we'll skip that
          * part. */
+        native_file_info.dwFileAttributes = attributes;
```

We may need to do something similar for the variant that takes a handle, falling back to looking at the path if the call to `GetFileInformationByHandle` fails.

andrenth replied:
I've fixed the `attributes` usage. About the second part, do you mean something like this in `efile_read_handle_info`?

```
posix_errno = internal_read_link(w->handle, &path);
if(posix_errno != 0) {
    return posix_errno;
}
if(!GetFileInformationByHandle(w->handle, &native_file_info)) {
    if(is_path_root(&path)) {
        // fallback
        return efile_read_info(&path, follow_links, result);
    }
    return windows_to_posix_errno(GetLastError());
}
```

andrenth replied:
I'm a bit out of my depth here as path handling in windows is a mystery to me... Wouldn't doing it inline instead of just calling efile_read_info() require duplicating its code in efile_read_handle_info()?

andrenth replied:
By the way, I was wondering if it wouldn’t be better to just return the error if `GetFileInformationByHandle` fails. 

If we fallback to getting the file info by path, wouldn’t it be possible for the path root to be replaced by something else after the error but before the fallback call, leaving the caller with an unexpected result (I have this descriptor, but the file info I requested is not related to it)?

andrenth replied:
I went ahead and implemented the fallback in the last commit so it's done if you prefer to go with it.

jhogberg replied:
> If we fallback to getting the file info by path, wouldn’t it be possible for the path root to be replaced by something else after the error but before the fallback call, leaving the caller with an unexpected result (I have this descriptor, but the file info I requested is not related to it)?

A "root path" is something that explicitly points at a drive like `C:\`, a network share, or similar. Should `C:\` start to point at a different drive, e.g. if the user unplugs a USB drive and then puts in a new one, all our open handles on that device will be dead and we'll error out when trying to figure out which path the handle refers to.

> By the way, I was wondering if it wouldn’t be better to just return the error if `GetFileInformationByHandle` fails.
> 

That would make it impossible to query root paths, which would differ from `efile_read_info`.

> I'm a bit out of my depth here as path handling in windows is a mystery to me... Wouldn't doing it inline instead of just calling efile_read_info() require duplicating its code in efile_read_handle_info()?

Sorry about the back-and-forth on this; after sleeping on it it's best if you call `efile_read_info` on failure. I was thinking about just saying "it's a directory" if we have a root path and leaving it at that, but then again there's bound to be some nasty edge cases around here. I'm not comfortable with Windows path handling either. :(

andrenth replied:
> Sorry about the back-and-forth on this; after sleeping on it it's best if you call efile_read_info on failure.

No problem! I've pushed an implementation of the fallback to `efile_read_info` which does it only for path roots. Do you think it should be done for any `GetFileInformationByHandle` errors?

jhogberg replied:
> No problem! I've pushed an implementation of the fallback to `efile_read_info` which does it only for path roots. Do you think it should be done for any `GetFileInformationByHandle` errors?

Thanks!

I don't think we should retry on all failures, roots are just an ugly special case we need to handle.

jhogberg replied:
We're still passing the path to `internal_read_link` in this function, causing `read_file_info` on links to fail.

`posix_errno = internal_read_link(path, &resolved_path);`

andrenth replied:
I’m really sorry about that. I’ll fix it when I arrive at work and do a full test run instead of relying just on reviewing the patches.

andrenth replied:
So, I wasn't able to run the automated tests on windows. I got an error when running the `Install` program saying that `-minimal` wasn't a recognized flag, so I edited the code to force the minimal mode without that flag, but then I got an error saying the `Install.ini` file could not be opened, although it's clearly there... So I'm not sure if something's broken in my Cygwin install.

I have tested manually, though, and I think things are working fine. I've tested links (via `ln -s` on Cygwin, shortcuts created on Windows Explorer, and the `mklink` command). I didn't test `\\localhost\c$` directly because this branch doesn't have the patch that allows `file:open` on directories. I did open a file using that kind of path (i.e. `\\localhost\c$\foo.txt`) and it works, though I'm not sure if that counts as the same test.

jhogberg replied:
> I’m really sorry about that. I’ll fix it when I arrive at work and do a full test run instead of relying just on reviewing the patches.

Don't sweat it, we're just glad to have your help. :o)

> So, I wasn't able to run the automated tests on windows. I got an error when running the `Install` program saying that `-minimal` wasn't a recognized flag, so I edited the code to force the minimal mode without that flag, but then I got an error saying the `Install.ini` file could not be opened, although it's clearly there... So I'm not sure if something's broken in my Cygwin install.

I usually run the tests in the source tree, maybe you'll have better luck with that;

```
%% in bash.
export TESTROOT=$ERL_TOP/release_tests;
make release_tests;
cd $TESTROOT/test_server;
$ERL_TOP/bin/werl;

%% in the erlang shell
ts:install().
ts:run(kernel, file_SUITE, [batch]).
```

> I have tested manually, though, and I think things are working fine. I've tested links (via `ln -s` on Cygwin, shortcuts created on Windows Explorer, and the `mklink` command). I didn't test `\\localhost\c$` directly because this branch doesn't have the patch that allows `file:open` on directories.

Feel free to squash and rebase to `master`.

> I did open a file using that kind of path (i.e. `\\localhost\c$\foo.txt`) and it works, though I'm not sure if that counts as the same test.

Sadly it doesn't, `\\localhost\c$\foo.txt` points at a file on a network share, but we want to query the share itself (`\\localhost\c$`) as if it were a directory, even though it's not.

Looking more closely at the tests it seems we're not checking this even though we should, can you add that path to `file_info_basic_directory` / `file_handle_info_basic_directory`?

andrenth replied:
Awesome, that worked! The commits are rebased to master and squashed, and the test cases for `\\localhost\c$` were added.

jhogberg replied:
Thanks, the tests ran well last night, so I'll merge this early next week if nothing new pops up. :)

jhogberg replied:
... `file_name_SUITE:normal` went wrong on some of our machines, didn't notice it until now :(

It crashes on the first line of `rm_r` with `{error, enoent}`, it looks like there's a problem with `read_link_info`.

andrenth replied:
Was that on Windows only?

andrenth replied:
Hi

Is there anything in common among the failing machines? Windows version, 32 bit vs 64 arch, etc? I can't reproduce the error here on Cygwin on 64-bit Windows 10 :(

andrenth commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -941,19 +972,10 @@ posix_errno_t efile_set_time(const efile_path_t *path, Sint64 a_time, Sint64 m_t
-        return windows_to_posix_errno(last_error);
-    }
-
     required_length = GetFinalPathNameByHandleW(link_handle, NULL, 0, 0);
```
Fixed.

andrenth commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -716,39 +716,90 @@ static int is_executable_file(const efile_path_t *path) {
-static int is_name_surrogate(const efile_path_t *path) {
-    HANDLE handle;
-    int result;
+static int is_handle_surrogate(HANDLE handle) {
```
I believe I've fixed this. I'll restart my test VM to ensure the commit works.

andrenth replied:
Seems to be working fine. I think the only thing pending is the question above w.r.t reusing efile_read_info as a fallback.

jhogberg commented on lib/kernel/doc/src/file.xml:
```diff
@@ -1438,7 +1438,10 @@ f.txt:  {person, "kalle", 25}.
-           </seealso></p>
+           </seealso></p>. This option has no effect when the function is
+           given an I/O device instead of a file name. Use
+           <seealso marker="#open/2"><c>open/2</c> with the <c>raw</c> mode to
```
Can you fix this in your branch as well?

andrenth replied:
It's already fixed: https://github.com/andrenth/otp/blob/f51535e77c17c481b1ae5ba8e46f040b9dd3d007/lib/kernel/doc/src/file.xml#L1442

I fixed it in a previous rebase and forgot to mention it.

jhogberg replied:
I thought so too until I tried adding it to our daily tests again. The `seealso` tag is fixed but there's still a trailing `.` in `</seealso></p>.`, it ought to be `</seealso>.</p>`.

You can test this yourself by moving to `lib/kernel/doc/src` and running `make xmllint`

andrenth replied:
Huh, I triple-checked and didn't see that. Fixed.

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -898,9 +898,8 @@ posix_errno_t efile_read_handle_info(efile_data_t *d, efile_fileinfo_t *result)
         build_file_info_times(&native_file_info, result);
     } else if(is_path_root(path)) {
```
This and the call to `efile_read_info` ought to take `&path` instead.

andrenth replied:
Fixed, also for `is_path_root`.

jhogberg commented:
I’m home with a cold and am having trouble checking our tests, but IIRC it failed on both 32 and 64-bit, and both Win7 and Win10. I think the ones that failed ran with a user that had permission to create links.

andrenth commented:
Hope you're feeling better... Only today I had some time to look into this again, and I managed to reproduce it when running the cywing console as administrator. Going to try to debug it...

andrenth commented:
So on windows, I've noticed that if I change some file and run `./otp_build boot -a` or `./otp_build boot update_preloaded --no-commit` again, it invariably fails with a strange error saying that a file doesn't exist, while it in fact does. So I end up having to switch back to master, reset all changes, and run the `autoconf`, `configure`, `boot` and `update_preloaded` steps again, then switch back to my branch and then do everything again. On VirtualBox, this takes a loooong time :(

Do you know of a way to make the recompile cycle shorter?

jhogberg commented:
That sounds really annoying, can you try `make emulator` instead of `./otp_build boot`?

andrenth commented:
Thanks, that was an incredible time saver!

I believe the latest commit fixes the problem, which was opening links without `FILE_FLAG_OPEN_REPARSE_POINT`, which resulted in the link destination being opened.

jhogberg commented on erts/emulator/nifs/win32/win_prim_file.c:
```diff
@@ -866,20 +866,22 @@ posix_errno_t efile_read_info(const efile_path_t *path, int follow_links, efile_
+        flags = FILE_FLAG_BACKUP_SEMANTICS;
+        if(attributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+            flags |= FILE_FLAG_OPEN_REPARSE_POINT;
+            is_link = is_name_surrogate(path);
```
I think we only want to open the reparse point if `is_link && !follow_links`, we don't want to open OneDrive markers for example.

andrenth replied:
Ok, I've made this change, and the tests still pass for me. I thought reparse points were just links but I guess it's more involved than that.

jhogberg replied:
Thanks, I've updated our tests.

jhogberg commented:
The tests ran fine over the weekend, so feel free to squash it. I'll merge it in the next few days. :)

andrenth commented:
Done. This was a though one :)

jhogberg commented:
Merged as 924cd70f8b7cf1fa2256055af39723b24fd6238e. Thanks again for the PR!
